File: c#/2751.cs
using System;

class Program
{
    static void Main(string[] args)
    {
        int N = int.Parse(Console.ReadLine());

        int[] Arr = Enumerable.Range(0, N)
            .Select(_ => int.Parse(Console.ReadLine()))
            .ToArray();

        // QuickSort(Arr, 0, N - 1);
        MergeSort(Arr, 0, N - 1);

        Console.WriteLine(string.Join("\n", Arr));
    }

    static Random rand = new();

    static void QuickSort(int[] Arr, int Left, int Right)
    {
        if (Left >= Right)
        {
            return;
        }

        var (i, j) = (Left, Right);
        int pivot = Arr[rand.Next(i, j + 1)]; // O(n^2)을 피할 수 있을지?

        while (i <= j)
        {
            while (Arr[i] < pivot)
            {
                i++;
            }
            while (Arr[j] > pivot)
            {
                j--;
            }

            if (i > j)
            {
                break;
            }

            (Arr[i], Arr[j]) = (Arr[j], Arr[i]);
            i++;
            j--;
        }

        QuickSort(Arr, Left, j);
        QuickSort(Arr, i, Right);
    }

    static void MergeSort(int[] Arr, int Left, int Right)
    {
        int Len = Right - Left + 1;

        if (Len <= 1)
        {
            return;
        }

        int Pivot = (Left + Right) / 2;

        MergeSort(Arr, Left, Pivot);
        MergeSort(Arr, Pivot + 1, Right);

        var (a, b) = (Left, Pivot + 1);
        int[] temp = new int[Len];

        for (int i = 0; i < Len; i++)
        {
            if (a < Pivot + 1 && b < Right + 1)
            {
                if (Arr[a] < Arr[b])
                {
                    temp[i] = Arr[a];
                    a++;
                }
                else
                {
                    temp[i] = Arr[b];
                    b++;
                }
            }
            else
            {
                if (a == Pivot + 1)
                {
                    temp[i] = Arr[b];
                    b++;
                }
                else
                {
                    temp[i] = Arr[a];
                    a++;
                }
            }
        }

        Array.Copy(temp, Arr, Len);
    }
}

File: python/2751.py
import sys
from random import randint


def main():
    N = int(sys.stdin.readline())
    arr = [int(sys.stdin.readline()) for _ in range(N)]

    # arr = quick_sort(arr)
    arr = merge_sort(arr)

    print(*arr, sep="\n")


def quick_sort(arr):
    if (length := len(arr)) <= 1:
        return arr

    pivot = arr[randint(0, length - 1)]
    i = 0
    j = length - 1

    while i <= j:
        while arr[i] < pivot:
            i += 1

        while arr[j] > pivot:
            j -= 1

        if i > j:
            break

        arr[i], arr[j] = arr[j], arr[i]
        i += 1
        j -= 1

    arr[: j + 1] = quick_sort(arr[: j + 1])
    arr[i:] = quick_sort(arr[i:])

    return arr


def merge_sort(arr):
    if (length := len(arr)) <= 1:
        return arr

    pivot = length // 2

    arr[:pivot] = merge_sort(arr[:pivot])
    arr[pivot:] = merge_sort(arr[pivot:])

    a, b = 0, pivot
    temp = []

    for _ in range(length):
        if a < pivot and b < length:
            if arr[a] < arr[b]:
                temp.append(arr[a])
                a += 1
            else:
                temp.append(arr[b])
                b += 1
        else:
            if a == pivot:
                temp.append(arr[b])
                b += 1
            else:
                temp.append(arr[a])
                a += 1

    return temp


if __name__ == "__main__":
    main()
